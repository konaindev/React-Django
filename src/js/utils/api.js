import axios from "axios";
import { store } from "../App";

const validateStatus = status => {
  return status >= 200 && status < 500;
};

export function axiosPost(url, data, headers = {}) {
  const { token } = store.getState();
  const { access } = token;

  const config = {
    method: "post",
    headers: { ...headers },
    data,
    url,
    validateStatus
  };
  if (access) {
    config.headers["Authorization"] = `bearer ${access}`;
  }
  if (data.toString() === "[object FormData]") {
    config.headers["content-type"] = "multipart/form-data";
  }
  // if (csrfProtect) {
  //   config.headers["X-CSRFToken"] = getCSRFToken();
  // }
  return axios(config);
}
export function axiosGet(url, config = { validateStatus }) {
  const { token } = store.getState();
  const { access } = token;
  const params = {
    method: "get",
    headers: { Accept: "application/json" },
    url,
    ...config
  };
  if (access) {
    params.headers["Authorization"] = `Bearer ${access}`;
  }
  return axios(params);
}
