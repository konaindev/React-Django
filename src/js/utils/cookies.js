/**
 * @description Return the value of a named cookie.
 *
 * Returns an empty string if the cookie cannot be identified.
 *
 * @returns {string}
 */
export const getCookie = name => {
  const cookieName = `${name}=`;
  const decoded = decodeURIComponent(document.cookie);
  const splits = decoded.split(";");
  for (const split of splits) {
    const trimmed = split.trim();
    if (trimmed.indexOf(cookieName) == 0) {
      return trimmed.substring(cookieName.length, trimmed.length);
    }
  }
  return "";
};

export const setCookie = (name, value) => {
  document.cookie = `${name}=${value}`;
};
