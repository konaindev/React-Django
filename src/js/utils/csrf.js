/**
 * @description Gets the current CSRF token.
 *
 * Returns null if none can be found.
 *
 * @returns {string=}
 */
export const getCSRFToken = () =>
  document.querySelector("[name=csrfmiddlewaretoken]").value;
