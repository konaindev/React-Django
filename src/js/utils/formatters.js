import React from "react";

/**
 * @description Return the BCP language identifier for the current user
 */
const getLanguage = () =>
  navigator.language ||
  navigator.browserLanguage ||
  (navigator.languages || ["en-US"])[0];

/**
 * @description Convert a value (like 0.25) to a display string (25%)
 *
 * @param {number|string} value A percentage value out of 1.0 to format
 * @param {number} decimals The number of decimal places to include
 */
export const formatPercent = (value, maxFractionDigits = 0, minFractionDigits) => {

  if (minFractionDigits === undefined) {
    minFractionDigits = maxFractionDigits;
  }

  const formatter = Intl.NumberFormat(getLanguage(), {
    style: "percent",
    minimumFractionDigits: minFractionDigits,
    maximumFractionDigits: maxFractionDigits
  });
  return formatter.format(value);
};

/**
 * @description Convert a value (like 4500) to a display string (4,500)
 *
 * @param {number|string} value A numerical value to format
 * @param {number} decimals The number of decimal places to include
 */
export const formatNumber = (value, decimals = 0) => {
  const formatter = Intl.NumberFormat(getLanguage(), {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals
  });
  return formatter.format(value);
};

/**
 * @description Convert a value (like 4500) to a USD display string ($4,500)
 *
 * @param {number|string} value A numberical value to format
 * @param {boolean} cents If true, show cents
 * @param {string} currency The ISO 4217 currency code
 */
export const formatCurrency = (value, cents = false, currency = "USD") => {
  const formatter = Intl.NumberFormat(getLanguage(), {
    style: "currency",
    minimumFractionDigits: cents ? 2 : 0,
    maximumFractionDigits: cents ? 2 : 0,
    currency
  });
  return formatter.format(value);
};

/**
 * @description Convert a value (like 4500) to a shorthand USD display string ($4.5k)
 *
 * @note This is always formatted in the en-US locale, because otherwise my hacknology could break.
 *
 * @param {number|string} value A numerical value to format
 * @param {string} currency The ISO 4217 currency code
 */
export const formatCurrencyShorthand = (value, currency = "USD") => {
  const levels = ["", "k", "m", "b"];

  // reduce value down to dollars, thousands, or millions of dollars
  let number = Number(value);
  let levelIndex = 0;
  while (Math.abs(number) >= 1000) {
    number = number / 1000.0;
    levelIndex += 1;
  }
  const level = levels[levelIndex];

  const maximumDigits = Math.abs(number) < 1 ? 1 : 0;

  // format the result as a dollar value
  const formatter = Intl.NumberFormat("en-US", {
    style: "currency",
    minimumFractionDigits: 0,
    maximumFractionDigits: maximumDigits,
    useGrouping: false,
    currency
  });

  // return the result with appropriate dollar value
  return `${formatter.format(number)}${level}`;
};

/**
 * @description Format a date like it's the USA!
 */
export const formatDate = (value, year = true) => {
  const options = { month: "numeric", day: "numeric" };
  if (year) {
    options.year = "2-digit";
  }
  const formatter = Intl.DateTimeFormat("en-US", options);

  // I *loathe* Javascript dates. I never get things right the first time.
  // This is a fix:
  const rawDate = new Date(value);
  const timezoneOffset = rawDate.getTimezoneOffset() * 60000;
  const d = new Date(rawDate.getTime() + timezoneOffset);

  return formatter.format(d);
};

/**
 * @description Format a multiple by putting the letter x next to it.
 *
 * @param {number|string} value A value to format
 *
 * @note This could probably use some smarts.
 */
export const formatMultiple = value => {
  return `${value}x`;
};

/**
 * @description Format the difference between two percentages as "points"
 *
 * @param {number|string} value A number of points, divided by 100, to format
 *
 * @note The `value` parameter is intended to be
 */
export const formatDeltaPercent = value => {
  const number = Number(value) * 100;
  const digits = Math.abs(number) < 1 ? 1 : 0;
  return `${formatNumber(number, digits)}pts`;
};

// XXX move this somewhere else @FIXME since it implicitly depends on React

/**
 * @description Wrap a value formatter to properly format "Target: " strings.
 *
 * @note If the underlying target value is null, we return an empty string.
 */
export const targetFormatter = formatter => targetValue =>
  targetValue == null ? (
    <span>&nbsp;</span>
  ) : (
    `Target: ${formatter(targetValue)}`
  );

export const formatTargetPercent = targetFormatter(formatPercent);
