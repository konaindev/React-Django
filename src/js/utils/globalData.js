/**
 * @file Simple tools for getting page-wide ("global") data values.
 */

/**
 * @description Extract a javascript object from the page.
 *
 * The provided id should be the identity of a <script> tag on the page
 * that contains "global" data in the form of a JSON structure.
 *
 * These values can be created using Django 2.1's json_script tag.
 */
export const getGlobalData = id => {
  const element = document.getElementById(id);
  const data = element ? JSON.parse(element.textContent) : {};
  return data;
};
