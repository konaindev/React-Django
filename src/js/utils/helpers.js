export const isNil = value => typeof value === "undefined" || value === null;
