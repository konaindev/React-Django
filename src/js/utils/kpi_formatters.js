import {
  configuredFormatCurrency,
  configuredFormatNumber,
  configuredFormatPercent
} from "./formatters";

const KPI_FORMAT = {
  // Investment
  acq_reputation_building: configuredFormatCurrency(),
  acq_demand_creation: configuredFormatCurrency(),
  acq_leasing_enablement: configuredFormatCurrency(),
  acq_market_intelligence: configuredFormatCurrency(),
  ret_reputation_building: configuredFormatCurrency(),
  ret_demand_creation: configuredFormatCurrency(),
  ret_leasing_enablement: configuredFormatCurrency(),
  ret_market_intelligence: configuredFormatCurrency(),

  // Funnel Volumes
  usvs: configuredFormatNumber(),
  inquiries: configuredFormatNumber(),
  tours: configuredFormatNumber(),
  lease_applications: configuredFormatNumber(),
  leases_executed: configuredFormatNumber(),

  // Leasing
  occupiable_units: configuredFormatNumber(),
  leased_units: configuredFormatNumber(),
  leased_rate: configuredFormatPercent(),
  renewal_rate: configuredFormatPercent(),
  lease_cd_rate: configuredFormatPercent(),
  lease_renewal_notices: configuredFormatNumber(),
  lease_renewals: configuredFormatNumber(),
  lease_vacation_notices: configuredFormatNumber(),

  // Activity
  move_outs: configuredFormatNumber(),
  move_ins: configuredFormatNumber(),
  occupancy_rate: configuredFormatPercent(),
  occupied_units: configuredFormatNumber(),

  // Investment
  acq_investment: configuredFormatCurrency(),
  ret_investment: configuredFormatCurrency(),
  investment: configuredFormatCurrency(),

  // Revenue
  estimated_acq_revenue_gain: configuredFormatCurrency(),
  estimated_ret_revenue_gain: configuredFormatCurrency(),
  estimated_revenue_gain: configuredFormatCurrency(),

  // ROMI
  romi: configuredFormatNumber(),
  ret_romi: configuredFormatNumber(),
  acq_romi: configuredFormatNumber(),

  // Rent
  exe_to_lowest_rent: configuredFormatPercent(),

  // Funnel Conversions
  usv_exe: configuredFormatPercent(),
  usv_inq: configuredFormatPercent(),
  inq_tou: configuredFormatPercent(),
  tou_app: configuredFormatPercent(),
  app_exe: configuredFormatPercent(),

  // Cost Pers
  usv_cost: configuredFormatCurrency(),
  inq_cost: configuredFormatCurrency(),
  tou_cost: configuredFormatCurrency(),
  app_cost: configuredFormatCurrency(),
  exe_cost: configuredFormatCurrency()
};

export function formatKPI(name, value) {
  if (!KPI_FORMAT[name]) {
    throw new Error("Key provided to formatKPI is invalid");
  }
  return KPI_FORMAT[name](value);
}
