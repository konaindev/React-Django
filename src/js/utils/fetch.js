import { getCSRFToken } from "./csrf";

export function post(url, data, headers = {}, csrfProtect = true) {
  const init = {
    method: "POST",
    body: data,
    headers: new Headers(headers)
  };
  if (csrfProtect) {
    init.headers.append("X-CSRFToken", getCSRFToken());
  }
  return fetch(url, init);
}
