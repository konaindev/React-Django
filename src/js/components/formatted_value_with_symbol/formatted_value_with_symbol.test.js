import FormattedValueWithSymbol from "./index";
import renderer from "react-test-renderer";
import { formatNumber, formatCurrencyShorthand } from "../../utils/formatters";

describe("FormattedValueWithSymbol", () => {
  it("renders default correctly", () => {
    const props = {
      value: "495200.00",
      formatter: formatCurrencyShorthand
    };

    const tree = renderer
      .create(<FormattedValueWithSymbol {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders plus/minus sign correctly", () => {
    const props = {
      value: "354000.00",
      formatter: formatCurrencyShorthand,
      symbolType: "sign"
    };

    const tree = renderer
      .create(<FormattedValueWithSymbol {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders multiple-x correctly", () => {
    const props = {
      value: -34,
      formatter: formatNumber,
      symbolType: "multiple"
    };

    const tree = renderer
      .create(<FormattedValueWithSymbol {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
