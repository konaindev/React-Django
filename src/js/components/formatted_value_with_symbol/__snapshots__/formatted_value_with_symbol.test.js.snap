// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedValueWithSymbol renders default correctly 1`] = `
<span
  className="formatted-value-with-symbol"
>
  $495k
</span>
`;

exports[`FormattedValueWithSymbol renders multiple-x correctly 1`] = `
<span
  className="formatted-value-with-symbol"
>
  <span
    className="formatted-value-with-symbol__minus"
  >
    -
  </span>
  34
  <span
    className="formatted-value-with-symbol__times"
  >
    x
  </span>
</span>
`;

exports[`FormattedValueWithSymbol renders plus/minus sign correctly 1`] = `
<span
  className="formatted-value-with-symbol"
>
  <span
    className="formatted-value-with-symbol__plus"
  >
    +
  </span>
  $354k
</span>
`;
